### ISSUE TITLE
Public interface for Keyframe

### ISSUE THREAD

**Issue by musteresel** [OpenShot/libopenshot#370]
The current public interface of `Keyframe` includes the public member `Points`.  **I'm about to remove this member.  This <s>will</s>** might **break a lot of code** (<s>basically every place where Python code - e.g. in OpenShot-qt - uses `keyframe["Points"]` ... which is currently 88 places in OpenShot-qt</s> *nothing in OpenShot-qt as far as I can see now*).

There's `AddPoint()`, `RemovePoint()`, `GetPoint()` among others.  These should be used instead of direct access to the `Points` member to give `Keyframe` a chance to keep it's internal state in order (e.g. sort order of points, interpolated values, ..).

All in all the public interface of `Keyframe` is a bit chaotic. It supports access to the points similar like a container (add, remove, find, ..) but of course also access to interpolated values (it's main job, after all).

The question is: How should the public interface of `Keyframe` look on the C++ side, and what should it look like on the Python side?

*edit: The question shifts a bit ... it's still important to get a clean public interface. But for openshot-qt the only really important functions are `Keyframe::SetJsonValue()`, `Keyframe::AddPoint`, the "get functions" and `Keyframe::Json()`.*

**Comment by musteresel** [comment]
OpenShot-qt uses `Points` mostly in these ways:

 - iterating over it and indexing into it, giving read **and write** access to each point
 - `append`ing new points
 - `remove`ing points by point value (not index)

This is a huge problem. The python code currently does stuff like this ([src/windows/views/timeline_webview.py#L334](https://github.com/OpenShot/openshot-qt/blob/develop/src/windows/views/timeline_webview.py#L334))

~~~python
brightness["Points"][-1]["co"]["X"] = round(duration * fps_float) + 1
~~~

Removing `Points` and providing only `AddPoint` and `RemovePoint` makes this much more complicated to write for the python side.

One possible approach is the following:

`Keyframe` (C++ side) can provide access to a proxy class which acts like a container of points. E.g. you can iterate over points, remove and add points, and access individual points by index. Instead of giving access to the points however, the proxy container gives access to "proxy points".  These points notify the `Keyframe` instance the point is part of over any changes that are made to them.

Example:

~~~python
p = some_keyframe.points()[20]
p = some_other_point
~~~

This would call `ProxyPoint::operator=` on the C++ side, which in turn would call `RemovePoint()` for the point which is proxied, then `AddPoint()` for the given new point and set itself up as proxy for the new point (which might have another index!)

Thoughts on this?

**Comment by SuslikV** [comment]
Usually, libs modified in the way that didn't broke functionality of applications that relays on it. Have no other thoughts.

**Comment by musteresel** [comment]
**`GetLength()`**: This currently provides access to `Values.size()`.  I used it (instead of `Values.size()`) to make `Values` private.  It's only used in checks like: `some_keyframe.GetLength() > 1` (in current develop these look like `some_keyframe.Values.size() > 1`).

It's behavior isn't like you'd expect "length" to behave:

~~~python
>>> f = openshot.Keyframe()
>>> f.GetLength()
0 # Values is empty
>>> f.AddPoint(1,1)
>>> f.GetLength()
1 # Values with a single point at X = 1
>>> f.AddPoint(2,1)
>>> f.GetLength()
3 # Values with points at X = 0, 1, 2
>>> f.AddPoint(200,1)
>>> f.GetLength()
201 # Values with points at X = 0, 1, ..., 200
>>> g = openshot.Keyframe()
>>> g.AddPoint(200,1)
>>> g.GetLength()
1 # Just a single point at X = 200 in Values
>>> g.AddPoint(1,1)
>>> g.GetLength()
201 # Values with points at X = 0, 1, ..., 200
~~~

What these tests seem to be about is whether there's any interpolation happening in the keyframe (happens when more than 2 points are there, then there are also more than 1 entries in `Values`).
**So I suggest to provide and use a function `bool IsConstant()` which returns `true` if the `Keyframe` has no points, just a single point, or just points with the same Y values.**

This function can then also cover cases in the code which access `Points.size()`: The test there look always like `some_keyframe.Points.size() > 1` which is equivalent to asking "is there interpolation happening?". This changes calling code (in libopenshot) from:

 - `if (some_keyframe.Values.size() > 1)` to `if (! some_keyframe.IsConstant())`
 - `if (some_keyframe.Points.size() > 1)` to `if (! some_keyframe.IsConstant())`

According to `find openshot-qt/ -type f -name '*.py' -exec grep -FHn GetLength '{}' ';'` OpenShot-qt doesn't use this function.

`GetLength()` might still be useful to know the last (largest) X coordinate which is an interpolation; it should then get a better name, though?

**Comment by musteresel** [comment]
From @ferdnyc in https://github.com/OpenShot/libopenshot/issues/373#issuecomment-558135687, emphasis mine:

> See... I'm glad you brought this up, actually, because **populating the clip data with a set of initial parameters doesn't mean you don't still need default values** — especially if the user has access to modify the values from their initial state. And it's a pretty big design flaw in libopenshot's current properties system, that it doesn't have any support for applying default values to the properties.
> [..]
> **Without defaults, there's no way to return the data to an unmodified state** — clearing a property zeroes it, rather than resetting it, which isn't the same thing. Zero happens to be the default for a few of the properties, but only by accident. And zeroing any of the rest just breaks them.
> [..]
> Implementation-wise, maybe that means inserting a keyframe somewhere the user can't modify it, like at "frame 0". (hey, maybe it's a lucky thing after all, that we start from frame 1!) That's just behind-the-scenes details. But however it's accomplished, there needs to be an **immutable baseline state that's outside of the user's control**, a state the data will return to when they remove their modifications. We haven't got truly non-destructive editing without it.

These are very good points!

*In the following I assume #374 get's merged, so I'm talking about the code at commit a67fb9555cc4bd84f95b1c2ce0f6144c25621ccd*

Currently we have:
 * `Keyframe::Keyframe()` which initializes the Keyframe with no points.
 * `Keyframe::Keyframe(double val)` which initializes the Keyframe with a single point at `(1,  val)`
 * `Keyframe::GetValue(int64_t)` which returns `0` if there are no points, and if there's just one point it returns that points Y value.

This allows us to "set defaults" via (e.g. in `Clip`):

~~~c++
	// Init scale curves
	scale_x = Keyframe(1.0);
	scale_y = Keyframe(1.0);

	// Init location curves
	location_x = Keyframe(0.0);
	location_y = Keyframe(0.0);

	// Init alpha
	alpha = Keyframe(1.0);
~~~

As identified by @ferdnyc above these values are not immutable, though. The user is able to delete that one point added by `Keyframe::Keyframe(double)`.

I propose the following change:

* `Keyframe::Keyframe(double val)` set's a new member `default_value`. It does **not** add any points to the Keyframe
* `Keyframe::Keyframe()` just calls `Keyframe::Keyframe(0)`
* `Keyframe::GetValue(int64_t)` returns `default_value` if there are no points in the Keyframe.
* `default_value` won't be changed by any member function.

This means code like the excerpt above from `Clip` should continue as expected. The only difference is that the user cannot delete that "default value point". If the user adds a point to the keyframe, then they basically just set a new "global constant" value to the Keyframe. Only once they add another point real interpolation is used.

I think this solves the issue with minimal implementation work?

**Comment by musteresel** [comment]
`Keyframe::Contains(Point)` ... this does **not**, as the name suggest, check if the given point is "contained" in the Keyframe. It tests whether there's a point at that given points X coordinate.
Thus I propose to change it to:

`Keyframe::HasPointAt(uint64_t) const`

Impact on openshot: None.

**Comment by musteresel** [comment]
`int64_t Keyframe::FindIndex(Point p) const`

For what do we need this function? It tries to find a point with the same X and Y coordinates like the given point. It does not take handles into account. Not used in openshot as far as I can see, only used in libopenshot from within Keyframe itself (easy to adjust).

I suggest to remove this and rather provide another way to get the index to a point.

Maybe together with: `GetClosesPoint(Point, bool)` ? This just uses the X coordinate of the point.

I suggest:

 `int64_t Keyframe::FindClosestPoint(int64_t x, bool) const` which returns the *index* (if found, else -1) of the closest point to the left / right.  (read only) access to the point is given via `const Point & Keyframe::GetPoint(int64_t)`.

`GetPreviousPoint` ... is not really used. I suggest removing it .. it can be implemented with a simple `FindClosesPoint(idx) - 1` (plus a check whether this is less than -1 to keep >= -1)

**Comment by musteresel** [comment]
`Point Keyframe::GetMaxPoint() const`

This function is actually used by libopenshot ... not by openshot-qt though:

~~~
libopenshot/src/QtImageReader.cpp:195:                  float max_scale_x = parent->scale_x.GetMaxPoint().co.Y;
libopenshot/src/QtImageReader.cpp:196:                  float max_scale_y = parent->scale_y.GetMaxPoint().co.Y;
libopenshot/src/QtImageReader.cpp:202:                  float max_scale_x = parent->scale_x.GetMaxPoint().co.Y;
libopenshot/src/QtImageReader.cpp:203:                  float max_scale_y = parent->scale_y.GetMaxPoint().co.Y;
libopenshot/src/FFmpegReader.cpp:1283:                          float max_scale_x = parent->scale_x.GetMaxPoint().co.Y;
libopenshot/src/FFmpegReader.cpp:1284:                          float max_scale_y = parent->scale_y.GetMaxPoint().co.Y;
libopenshot/src/FFmpegReader.cpp:1290:                          float max_scale_x = parent->scale_x.GetMaxPoint().co.Y;
libopenshot/src/FFmpegReader.cpp:1291:                          float max_scale_y = parent->scale_y.GetMaxPoint().co.Y;
~~~

As can be seen, it's only used for the Points Y coordinate. Yet I think it's a good idea to not only return the max value but also *where* that value is. But this should be done with a `Coordinate`, not a `Point`. We don't care about handles and interpolation here, after all. So a better name ... perhaps `GetCoordinateWithMaxY()`?

**Comment by musteresel** [comment]
`GetDelta(index) == 0` when `index < 1` or `index >= last_point.co.X` and `GetDelta(1) == first_point.co.Y`. This is very strange ... ?

I'd like to change this to:

~~~c++
double Keyframe::GetDelta(int64_t index) const
{
    return GetLong(index) - GetLong(index - 1);
}
~~~

I'm unsure about the impact on the only usage of `GetDelta` in both libopenshot and openshot-qt, though:

~~~
src/Clip.cpp:413:           int delta = int(round(time.GetDelta(frame_number)));
~~~

**Comment by musteresel** [comment]
~~~c++
void Keyframe::RemovePoint(Point p); // <<= let's get rid of this one!
void Keyframe::RemovePoint(int64_t index);
~~~

We can search for a specific point and then remove it by it's index, there's no need for that remove function with a point argument.

`void Keyframe::UpdatePoint(int64_t, Point)` ... as strange as this function is this should stay IMO. It's a good way to implement this workflow:

 1. Get some point by index
 2. Modify it's X coordinate.
 3. Update that point.

Perhaps a "Move point" function would be more appropriate .. but this creates a whole new set of difficulties.

**Comment by musteresel** [comment]
`void Keyframe::ScalePoints(double scale)`

**This is underspecified!** What happens with a scale < 0. What when scale == 0? What when the scale is so small that two points with prior different X coordinates land on the same? Which point should be used then? The later of course ... but we do not allow points with same X coordinates (see `AddPoint`). Just remove the point that was previously the one with the smaller X coordinate?

**Comment by musteresel** [comment]
Last but not least, **why do we use `double` for the X coordinates?** These are actually frame numbers, and we round them eitherway. Using an `int64_t` (or perhaps `int32_t`, or even unsigned types) makes much more sense.

**Comment by ferdnyc** [comment]
> Last but not least, **why do we use `double` for the X coordinates?** These are actually frame numbers, and we round them eitherway. Using an `int64_t` (or perhaps `int32_t`, or even unsigned types) makes much more sense.

I would be wary of changing that — there's a long-running (and, I suspect, never-ending) debate about whether keyframe positions should be stored as _frame numbers_ or as _time indexes_ (which would be expressed in terms of whole and fractional seconds).

Both have their advantages and disadvantages, and it's **unlikely** that we'd change horses _this_ far across the stream, but point is that the X coordinates always being integer frame numbers is not a design requirement. It's merely a decision made in the current implementation, and theoretically could change in the future. Supporting non-integer X coordinates affords that possibility.

**Comment by ferdnyc** [comment]
> I propose the following change:
> 
> * `Keyframe::Keyframe(double val)` set's a new member `default_value`. It does **not** add any points to the Keyframe
> * `Keyframe::Keyframe()` just calls `Keyframe::Keyframe(0)`
> * `Keyframe::GetValue(int64_t)` returns `default_value` if there are no points in the Keyframe.
> * `default_value` won't be changed by any member function.
> 
> This means code like the excerpt above from `Clip` should continue as expected. The only difference is that the user cannot delete that "default value point". If the user adds a point to the keyframe, then they basically just set a new "global constant" value to the Keyframe. Only once they add another point real interpolation is used.
> 
> I think this solves the issue with minimal implementation work?

I like where this is going, for sure... it's a reasonable-sounding proposal on the face of it, though it's pretty daunting to try and apply it to all of the myriad quirks in the current keyframe system, in order to (hopefully) puzzle out whether it covers them all. Especially in terms of what happens when you go from 0 to 1 point, from 1 to 2 points, from 2 to 1 point, or from 1 to 0 points, since OpenShot special-cases some of those conditions in confusing ways.

It seems OK to me, though if I'm interpreting correctly it _would_ alter at least some of the current functionality. Quite possibly for the **better**, honestly, since these are things that many users report finding confusing. But it would still be a change.

I'm talking about things like this, for example:
1. Right now if you position the playhead somewhere _other_ than a clip's first frame and make a `Scale X` adjustment, that creates the _second_ keyframe for that property (the first being the initial `1.0`) so it's created as a linear interpolation from the initial value. `Scale X` starts at `1.0`, smoothly fades to the new value over time, and then holds constant at that second value until the end of the clip.
   * It's possible there are users who rely on this behavior, but I suspect very few of them. It's almost never what you _want_.
   * Even if a user does want a property adjustment to occur somewhere after the beginning of the clip, they're most likely not going to want it to _slowly_ apply over the course of the entire clip. 
   * So they're still going to have to create a "start" keyframe point for the interpolation.
1. With this change, because the newly-added value would be the **first** keyframe, it would be applied over the entire length of the clip.
   * This might disconcert users who are used to the current functionality and expect that when they create keyframes at `X > 1` frame positions, the initial value won't be affected.
   * But it will better _match_ the expectations of the many, many other users who expect their property adjustments to apply over the entire clip duration regardless where the playhead was positioned when they did it.
1. Adding a **second** keyframe which also has a frame position `X > 1` will create a linear interpolation only _between_ the two, with constant values on either side. Which means that with this new code, same as the old, only one additional 'start' point has to be added for a mid-clip keyframe change.
   * This should win back the users familiar with the current behavior.
   * The second (really third) "start" keyframe point I said they'd end up having to create, back in point 1, is still the _only_ additional point they'd have to create with the new code.
   * That second (start-of-property-interpolation) point with a value of `1.0` will be sufficient for `Scale X` to have a constant `1.0` value on every frame _before_ the "start" keyframe, exactly like the current behavior.

I also _feel like_ this could actually allow for some real improvements in OpenShot's UI/UX:

1. Obviously, first and foremost, it would mean that deleting the lone keyframe for a Property would no longer be an inadvisable action, but it would simply represent the act of "resetting" the value. That's a **huge** win when it comes to user expectations and DWIM behavior.
1. It would allow us to add an _**explicit**_ reset action to the context menu for each Property, which would delete _all_ of the corresponding keyframe points, restoring the default value.
1. Currently, OpenShot special-cases single-point properties, and treats them as non-keyframed. That's done so that the Properties list doesn't show an interpolation status for every single keyframable property. But it gets confusing when you remove the initial point, but replace it with a single point at some **other** position. The keyframe is _there_, and shows a little green mark just like any other, but the green/blue highlighting of the Properties row is disabled, so there's no way to actually determine the location of that point by moving the Playhead.
    But with this change, that special-case logic for single points could be eliminated, so that the value is displayed as modified even if there's only a single keyframe point, and regardless whether it's located at the first frame or a later one. That would allow non-default values to be identified and located far more easily. (Not that it's always really _necessary_ to locate the lone keyframe point for a property. If the intent is to simply reset its value, "Remove keyframe" will remove the last point _no matter where_ it is in relation to the playhead. The new functionality would allow users to make better use of that behavior.)

**Comment by musteresel** [comment]
> alter at least some of the current functionality

Definitively. The new behavior would be as you described, and I think this would be a very intuitive way of working with the Keyframes.  Still it requires changes to openshot-qt ... which means that we need to make sure that openshot-qt uses the correct version of the libopenshot library. Otherwise things might fall apart.

Of course, since libopenshot exports its version it would *theoretically* be possible to have openshot-qt adapt to the available library version, e.g.

~~~python
if version_compare(libopenshot_version, "0.2.2") > 0:
    keyframe_behavior = new_keyframe_behavior
else:
    keyframe_behavior = old_keyframe_behavior
~~~
It's probably a lot of effort to bring "all" keyframe handling code into such a module in the first place, though.

> I also feel like this could actually allow for some real improvements in OpenShot's UI/UX:

@SuslikV In light of your PR https://github.com/OpenShot/openshot-qt/pull/3084 .. what do you think about this?

**Comment by musteresel** [comment]
@ferdnyc Do you see any issues with the other - non breaking - changes?

**Comment by musteresel** [comment]
> [..] whether keyframe positions should be stored as frame numbers or as time indexes (which would be expressed in terms of whole and fractional seconds).

Hm ... well, we have `GetValue(int64_t)`. Not `GetValue(double)`. We `round()` every X coordinate. So with that, we **are* using frame numbers. We only use `double` internally, and even there, we never use values with a fractional part. So, currently:

> X coordinates always being integer frame numbers is not a design requirement

We currently *need* to `round()` the X coordinates ... otherwise it get's **really** hard to e.g. work with an interpolation between very close values (then we need to handle division through zero, NaNs - not nice)

> It's merely a decision made in the current implementation, and theoretically could change in the future. Supporting non-integer X coordinates affords that possibility.

I guess the only thing that *might* need non-integer X coordinates is when we implement proper time scaling. So probably best to think about whether this change makes sense when we have an working / good approach to that.

**Comment by ferdnyc** [comment]
> Still it requires changes to openshot-qt ... which means that we need to make sure that openshot-qt uses the correct version of the libopenshot library. Otherwise things might fall apart.

That's easy enough, OpenShot already checks the libopenshot version against its own `info.MINIMUM_LIBOPENSHOT.VERSION` at startup, and refuses to run if the library version is below the minimum. So we just have to bump libopenshot's version, and raise the minimum value in OpenShot to match. 

I think it would be a mistake to even _try_ to provide compatibility with older library versions, since there's no reason anyone needs to run a newer OpenShot with an older libopenshot (or, no reason we have to _support_ it). In the active codebase, changes are made _**all the time**_ — well, "more often than I'd prefer" — that break compatibility between the latest OpenShot code and any libopenshot build that predates the most recent breaking commit. (Usually in ways that _can't_ be discerned by version number or other identifier, because those only get bumped on release, not for breaking changes. Something I have ranted about in the past, ineffectually and at length.<sup>OpenShot/openshot-qt#1371 and elsewhere</sup> )

#### Project data forward-compatibility

> Of course, since libopenshot exports its version it would _theoretically_ be possible to have openshot-qt adapt to the available library version, e.g.
> 
> ```python
> if version_compare(libopenshot_version, "0.2.2") > 0:
>     keyframe_behavior = new_keyframe_behavior
> else:
>     keyframe_behavior = old_keyframe_behavior
> ```

Checking the project data version, you mean? (Which is what `libopenshot_version` is, the value from the loaded data. As opposed to `openshot.GetVersion().ToString()`, aka `openshot.OPENSHOT_VERSION_FULL`, aka `openshot.Version` in the newest code — I've been slowly Pythonizing — which is the current runtime library version.)

We _could_ do that, and certainly project data compatibility is much more of a concern... but there too, I think we'd actually be just fine, and wouldn't even have to do anything special to deal with legacy project files in the new implementation.

Older project data will contain keyframe points `{ X: 1, Y: DEFAULT_VAL }` for every keyframable value — and those should function **exactly the same** in the new implementation as they did in the old one. The fact that the new implementation no longer _requires_ them, or that it will now function differently if they're _deleted_, is ultimately irrelevant. The existing project data should be handled exactly the same, at least until the user starts making further edits using the new version.

#### Project data _backwards_-compatibility
Going in the _other_ direction, loading newer project file data into an _older_ version of OpenShot... well, that's where things may get a little tricky. We haven't really progressed far enough along the data lifecycle to cover output data, but presumably that these changes would in turn result in changes to the project file contents (to eliminate the initial points containing default values), that would presumably make newer project files unloadable by earlier OpenShot releases.

For that reason, and because backwards compatibility is a valuable thing (especially when newer code is in flux, and may occasionally evidence buggy or unexpected behavior that forces users to go back to an earlier, known-functional version), it may be advisable at least in the short term to do a little extra work in the file-saving code to ensure that project files are written in a format that's compatible with earlier versions. That'd probably entail _inserting_ a first-frame keyframe point containing the default value for every property that doesn't already have a keyframe in that position.

We'd have to work out the exact methodology by which that's done so that the older code can interpret the data in a manner that's at least fairly close to how the newer version would interpret it. But the additional keyframes would be purely for the legacy code's benefit, the newer code would just ignore them.

#### Inserting new information into project data
I've experimented somewhat with this for other reasons, and if I'm recalling correctly I _**believe**_ that any random UN-expected data can be inserted into the project file's JSON dictionary objects without affecting the code — those keys will just be ignored. So, because keyframe points normally take one of these forms:
```json
{
    "co": {
        "X": 1.0,
        "Y": 1.0
    },
    "handle_left": {
        "X": 0.5,
        "Y": 1.0
    },
    "handle_right": {
        "X": 0.5,
        "Y": 0.0
    },
    "handle_type": 0,
    "interpolation": 0
}

{
    "co": {
        "X": 1.0,
        "Y": 1.0
    },
    "interpolation": 2
}
```
...we should be safe inserting a new key into either one, a boolean flag like `"compatibility": true` or `"default": true` or whatever, to indicate that the point is only there as a backwards-compatibility injection. The older code will ignore that key because it doesn't know what to do with it, and will use the point data the way it alway has. (IIRC it'll even _preserve_ the unknown key on re-save, which is kind of broken but handy for our purposes.) The _newer_ code can check for the flag and discard any points that have it set. (So, slight amendment to what I said about not having to do anything special when loading projects in the new code.)

But no need to even bother looking at the project data's `libopenshot_version` or `openshot_version`, since they doesn't tell us anything we can use: The _old_ code can't adjust for newer changes (by definition), and the _newer_ code doesn't have any reason to.

Then at some later point when it's no longer needed, we just drop out the compatibility code and stop inserting initial keyframe points.

**Comment by ferdnyc** [comment]
> `GetDelta(index) == 0` when `index < 1` or `index >= last_point.co.X` and `GetDelta(1) == first_point.co.Y`. This is very strange ... ?
> 
> I'd like to change this to:
> 
> ```c++
> double Keyframe::GetDelta(int64_t index) const
> {
>     return GetLong(index) - GetLong(index - 1);
> }
> ```

I was initially worried that you just suggested _removing_ this behavior:

> 1. Adding a second keyframe which also has a frame position X > 1 will create a linear interpolation only between the two, with constant values on either side. Which means that with this new code, same as the old, only one additional 'start' point has to be added for a mid-clip keyframe change.
>    * This should win back the users familiar with the current behavior.
>    * The second (really third) "start" keyframe point I said they'd end up having to create, back in point 1, is still the only additional point they'd have to create with the new code.
>    * That second (start-of-property-interpolation) point with a value of 1.0 will be sufficient for Scale X to have a constant 1.0 value on every frame before the "start" keyframe, exactly like the current behavior.

But if it's only used in this one place, for audio:
> ```
> src/Clip.cpp:413:           int delta = int(round(time.GetDelta(frame_number)));
> ```
...then it shouldn't matter. Though I honestly don't have a good understanding of how the audio processing works. I only know that it currently has bugs (`valgrind` shows invalid reads in every run, at the same places in the code), and that it's nebulous and confusing (which is why I stopped looking for the invalid reads).

Speaking of `GetDelta()`, though, I'm surprised you haven't brought up `IsIncreasing()` yet.

While the abundance of cut-and-paste code duplication is... one of libopenshot's (and OpenShot's) _charms_, of course, some things... `IsIncreasing()` is **line-for-line _identical_** to `GetDelta()` right up until they compute the return value, which would seem to leave it crying out to be replaced with a simple construct like `if (kf.GetDelta(f)  > 0) ...`.

(This is probably why you only found one use of `GetDelta()` in the code, also; there are four more `InIncreasings()` that should really be counted with it):
```grep
src/Clip.cpp:440: if (!time.IsIncreasing(frame_number))
src/Clip.cpp:494: 		if (!time.IsIncreasing(frame_number))
src/Clip.cpp:538: 		if (!time.IsIncreasing(frame_number))
src/Clip.cpp:582: 	if (!time.IsIncreasing(frame_number))
```

If `GetLong(index) - GetLong(index - 1)` would cover those cases as well, then I say put 'em _both_ on the scrap heap.

**Comment by ferdnyc** [comment]
> `IsIncreasing()` is **line-for-line _identical_** to `GetDelta()` right up until they compute the return value, which would seem to leave it crying out to be replaced with a simple construct, e.g.
> 
> ```c++
> if (kf.GetDelta(f)  > 0) {
>   // ...
> }
> ```

Actually, in the current code it would be `>= 0`, because as a comment notes:
https://github.com/OpenShot/libopenshot/blob/18811f40fea460a9bfd5f289c3a7d115dd977a7f/src/KeyFrame.cpp#L338-L339

`[citation needed]`

**Comment by musteresel** [comment]
> That's easy enough, OpenShot already checks the libopenshot version against its own `info.MINIMUM_LIBOPENSHOT.VERSION` at startup, and refuses to run if the library version is below the minimum. So we just have to bump libopenshot's version, and raise the minimum value in OpenShot to match.

Ok, that's good enough.

> **Project data forward-compatibility**
> [..]
> Older project data will contain keyframe points { X: 1, Y: DEFAULT_VAL } for every keyframable value — and those should function exactly the same in the new implementation as they did in the old one. The fact that the new implementation no longer requires them, or that it will now function differently if they're deleted, is ultimately irrelevant. The existing project data should be handled exactly the same, at least until the user starts making further edits using the new version.

Ok, so basically nothing to do here for libopenshot. It would get the JSON from the old project data and will basically generate the same JSON.  Makes it difficult to update Project data files in any way, but that's always the case and not really a libopenshot issue.

> **Project data backwards-compatibility**
> [..]
> For that reason, and because backwards compatibility is a valuable thing (especially when newer code is in flux, and may occasionally evidence buggy or unexpected behavior that forces users to go back to an earlier, known-functional version), it may be advisable at least in the short term to do a little extra work in the file-saving code to ensure that project files are written in a format that's compatible with earlier versions. That'd probably entail inserting a first-frame keyframe point containing the default value for every property that doesn't already have a keyframe in that position.
>
> We'd have to work out the exact methodology by which that's done so that the older code can interpret the data in a manner that's at least fairly close to how the newer version would interpret it. But the additional keyframes would be purely for the legacy code's benefit, the newer code would just ignore them.

Hm, good points. I haven't even thought about that yet.  Following the suggestions you've made, we come to something like this:

~~~c++
// new libopenshot
Keyframe kf(3.14)
kf.AddPoint(5,4);
kf.AddPoint(10,0);
kf.Json(); // Result below
~~~

~~~JSON
{ "points": [
  { # the special point
    "default": true
  },
  # the two added points
  ],
 "default_value": 3.14
}
~~~

The old libopenshot ignores the `"default"` and `"default_value"` keys and just works as it used to. The new code needs to look for the `"default_value"` key, and use that as default value. And the Point JSON parsing code needs to error out when it sees a `"default": true` (and Keyframe JSON parsing needs to catch and ignore that "error"). Good?

**Comment by musteresel** [comment]
> IsIncreasing() is line-for-line identical to GetDelta() right up until they compute the return value, which would seem to leave it crying out to be replaced with a simple construct, [..]
> [..]
> They use a literally IDENTICAL algorithm to arrive at the result!

No, that's not true. `IsIncreasing(index)` (in the old *and* new code) searches **forward** until the next value which is different from the value at `index`. If it's larger, then it returns `true`, otherwise `false`. If all values are equal (constant keyframe) .. then it still returns `true`.

The *old* `GetDelta()` was also performing this "search", but **backwards** (to previous values). And then it contained this:

~~~c++
		if (previous_repeats == 1)
			return current_value - previous_value;
		else
			return 0.0;
~~~

`previous_repeats` is 1 only when the immediately previous value does *not* repeat the value at the index. It's the number of previous values with the same value as the value at the index plus one.

So if the previous value is the same, it returns 0. Otherwise the difference to the previous value. Which is what the new code does with `GetValue(index) - GetValue(index - 1)`.

So the new code works exactly the same as the old, just faster :)

> But if it's only used in this one place, for audio: [..] ...then it shouldn't matter. 

The `GetDelta(i) == 0` for `i < 1` and `i >= last_x` special case code doesn't seem to be used by that audio processing code.  That just leaves the case for `GetDelta(1)` ... which returns the strange value. I'm looking what happens when I change this to return the more sensible 0 ...


**`IsIncreasing()`**

Well ... its behavior makes *halfway* sense. If the curve increases (even after being constant for some time) it returns `true`. If it decreases it returns `false`. If the curve *remains* constant it still returns `true`. But ...

~~~
src/Clip.cpp:440: if (!time.IsIncreasing(frame_number))
src/Clip.cpp:494: 		if (!time.IsIncreasing(frame_number))
src/Clip.cpp:538: 		if (!time.IsIncreasing(frame_number))
src/Clip.cpp:582: 	if (!time.IsIncreasing(frame_number))
~~~

... all these look like this:

~~~c++
if (!time.IsIncreasing(frame_number))
	reverse_buffer(samples);
~~~

**This doesn't make sense at all.**
I guess this needs to be fixed when approaching the whole time / speedup related issues.

**Comment by ferdnyc** [comment]
> > IsIncreasing() is line-for-line identical to GetDelta() right up until they compute the return value, which would seem to leave it crying out to be replaced with a simple construct, [..]
> > [..]
> > They use a literally IDENTICAL algorithm to arrive at the result!
> 
> No, that's not true. 

It is. Here they are side-by-side in Meld, from the current develop branch.

![image](https://user-images.githubusercontent.com/538020/69927366-c108b280-1485-11ea-9c0b-27ca082c3c67.png)

**Comment by ferdnyc** [comment]
From your comments it sounds like you've already fixed it so that it's no LONGER true, which is great because we still won't be running out of this kind of stuff anytime soon.

**Comment by ferdnyc** [comment]
Another thought I had... what about all of the other properties that _don't_ use Keyframes, and could still benefit from default values?

IOW, we've been focusing on Keyframe as the place to manage defaults, but is that _really_ the right approach? Or, even if it makes sense to create default values for Keyframe lists _as well_, should we still be adding default values to the properties system in general? (And if so, then what's the best way to manage the default values for properties that _are_ Keyframable?)

**Comment by SuslikV** [comment]
>... In light of your PR [OpenShot/openshot-qt#3084](https://github.com/OpenShot/openshot-qt/pull/3084) .. what do you think about this?

I need hours to read all this...
